clients/githubrepo/branches.go

```diff
@@ -266,5 +422,166 @@ func getBranchRefFrom(data *branch) *clients.BranchRef {
 		copyNonAdminSettings(rule, branchRule)
 	}
 
+	applyRepoRules(branchRef, rules)
+
 	return branchRef
 }
+
+func isPermissionsError(err error) bool {
+	return strings.Contains(err.Error(), "Resource not accessible by personal access token")
+}
+
+const (
+	ruleConditionDefaultBranch = "~DEFAULT_BRANCH"
+	ruleConditionAllBranches   = "~ALL"
+)
+
+func rulesMatchingBranch(rules []*repoRuleSet, name string, defaultRef bool) ([]*repoRuleSet, error) {
+	refName := fmt.Sprintf("refs/heads/%s", name)
+	ret := make([]*repoRuleSet, 0)
+nextRule:
+	for _, rule := range rules {
+		for _, cond := range rule.Conditions.RefName.Exclude {
+			if match, err := fnmatch.Match(cond, refName); err != nil {
+				return nil, fmt.Errorf("exclude match error: %w", err)
+			} else if match {
+				continue nextRule
+			}
+		}
+
+		for _, cond := range rule.Conditions.RefName.Include {
+			if cond == ruleConditionAllBranches {
+				ret = append(ret, rule)
+				break
+			}
+			if cond == ruleConditionDefaultBranch && defaultRef {
+				ret = append(ret, rule)
+				break
+			}
+
+			if match, err := fnmatch.Match(cond, refName); err != nil {
+				return nil, fmt.Errorf("include match error: %w", err)
+			} else if match {
+				ret = append(ret, rule)
+			}
+		}
+	}
+	return ret, nil
+}
+
+func applyRepoRules(branchRef *clients.BranchRef, rules []*repoRuleSet) {
+	for _, r := range rules {
+		var modded bool
+		for _, rule := range r.Rules.Nodes {
+			switch rule.Type {
+			case "DELETION":
+				if branchRef.BranchProtectionRule.AllowDeletions == nil || *branchRef.BranchProtectionRule.AllowDeletions {
+					branchRef.BranchProtectionRule.AllowDeletions = new(bool)
+					*branchRef.BranchProtectionRule.AllowDeletions = false
+					modded = true
+				}
+			case "NON_FAST_FORWARD":
+				if branchRef.BranchProtectionRule.AllowForcePushes == nil || *branchRef.BranchProtectionRule.AllowForcePushes {
+					branchRef.BranchProtectionRule.AllowForcePushes = new(bool)
+					*branchRef.BranchProtectionRule.AllowForcePushes = false
+					modded = true
+				}
+			case "PULL_REQUEST":
+				if applyPullRequestRepoRule(branchRef, rule) {
+					modded = true
+				}
+			case "REQUIRED_STATUS_CHECKS":
+				if applyRequiredStatusChecksRepoRule(branchRef, rule) {
+					modded = true
+				}
+			}
+		}
+		if modded {
+			adminEnforced := ruleAdminEnforced(r)
+			branchRef.BranchProtectionRule.EnforceAdmins = &adminEnforced
+		}
+	}
+}
+
+func applyPullRequestRepoRule(branchRef *clients.BranchRef, rule *repoRule) bool {
+	var modded bool
+	branchRules := branchRef.BranchProtectionRule.RequiredPullRequestReviews
+	dismissStale := readBoolPtr(rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush)
+	if dismissStale && !readBoolPtr(branchRules.DismissStaleReviews) {
+		branchRef.BranchProtectionRule.RequiredPullRequestReviews.DismissStaleReviews = &dismissStale
+		modded = true
+	}
```

Comment:
> things get weird when repository rules are layered

The more concerning bit is the value of `client.BranchRef.BranchProtectionRule.EnforceAdmins`.

The existing model assumes every property of the `BranchProtectionRule` is affected by the same admin bypass state. Once there are multiple repository rulesets that match the same branch, that is no longer true.
1. Create ruleset `restrict-deletes-with-bypass`
2. Create ruleset `require-pr-without-bypass`

What do you expect the value of `BranchProtectionRule.EnforceAdmins` to be? The way my draft is heading I think it depends on the order the rules were created 🙃 .
I'm wondering if to align with how repository rulests work, `BranchRef.BranchProtectionRule` should become a `[]BranchProtectionRule` slice? If so that's likely a large enough refactor for a separate PR.

Reply:
> The existing model assumes every property of the BranchProtectionRule is affected by the same admin bypass state. Once there are multiple repository rulesets that match the same branch, that is no longer true.
>
> 1. Create ruleset restrict-deletes-with-bypass
> 2. Create ruleset require-pr-without-bypass
>
> What do you expect the value of BranchProtectionRule.EnforceAdmins to be?

I think the simplest approach is maintaining the spirit of the existing check for now, so I would expect EnforceAdmin to be "do all rules have empty bypass actors", which would be `false`. A proper solution is likely out of scope for this PR.

There's also the aspect of the bypass list now being more than just admins. If you make a rule, but allow all team members/bots to bypass it, should the rule be counted? With a  unprivileged token, we can't really know.

Reply:
> A proper solution is likely out of scope for this PR.

It seems possible, just complicated. My approach was to:
1. Start from the `clients.BranchProtectionRule` from branch protection.
2. Loop through every matched `repoRuleSet`
   1. Translate the ruleset to an intermediate `clients.BranchProtectionRule`
   2. Compare the intermediate `clients.BranchProtectionRule` to the current base.  If it is more strict and `EnforceAdmins`, then "upgrade", if it is less strict then "downgrade".

Reply:
I like the intermediate rule approach, i think it simplifies the readability of parsing the repo rules.

I'm still not convinced on 2ii though. As you mention in your comments, it's tricky to determine if something is a downgrade and I think there's ordering aspects to consider too.